Fixes to ExitStatus and its docs

ijackson opened the pull request:
* On Unix, properly display every possible wait status (and don't panic on weird values)
* In the documentation, be clear and consistent about "exit status" vs "wait status".

rust-highfive commented:
r? @dtolnay

(rust-highfive has picked a reviewer for you, use r? to override)

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/1954918474) [(plain)](https://github.com/rust-lang/rust/commit/31e89075189918ca320f1689588d6fc0b1dc4839/checks/1954918474/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
Checking which error codes lack tests...
Found 436 error codes
Found 0 error codes with no tests
Done!
tidy error: `/checkout/library/std/src/sys/unix/process/process_unix.rs:546` contains `#[test]`; unit tests and benchmarks must be placed into separate files or directories named `tests.rs`, `benches.rs`, `tests` or `benches`
some tidy checks failed

command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/tidy" "/checkout" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "/checkout/obj/build"
expected success, got: exit code: 1

```

</details>

dtolnay reviewed (commented):
LGTM once the tidy failure is resolved. Thanks!

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/1957147564) [(plain)](https://github.com/rust-lang/rust/commit/a6b229eaaaa956837ea8c87e8d060b9589ef63a8/checks/1957147564/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
skip untracked path cpu-usage.csv during rustfmt invocations
skip untracked path src/doc/book/ during rustfmt invocations
skip untracked path src/doc/rust-by-example/ during rustfmt invocations
skip untracked path src/llvm-project/ during rustfmt invocations
Diff in /checkout/library/std/src/sys/unix/process/process_unix/tests.rs at line 2:
 fn exitstatus_display_tests() {
     // In practice this is the same on every Unix.
     // If some weird platform turns out to be different, and this test fails, use #[cfg].
-    use crate::process::ExitStatus;
     use crate::os::unix::process::ExitStatusExt;
+    use crate::process::ExitStatus;
 
     let t = |v, s| assert_eq!(s, format!("{}", <ExitStatus as ExitStatusExt>::from_raw(v)));
 
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2018" "--unstable-features" "--skip-children" "--check" "/checkout/library/std/src/sys/unix/process/process_unix.rs" "/checkout/library/std/src/sys/unix/mod.rs" "/checkout/library/std/src/sys/unix/process/process_unix/tests.rs" "/checkout/library/std/src/sys/unix/process/process_common/tests.rs" "/checkout/library/std/src/sys/unix/process/process_common.rs" "/checkout/library/std/src/sys/unix/process/mod.rs" "/checkout/library/std/src/sys/windows/pipe.rs" "/checkout/library/std/src/sys/unix/process/zircon.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
Build completed unsuccessfully in 0:00:16
```

</details>

ijackson commented:
> LGTM once the tidy failure is resolved. Thanks!

Oops. My privsep workflow means I sometimes forget the tidy check.  Fixed now (and squashed the fix in).  Thanks.

ijackson commented:
@dtolnay Hi.  Sorry to hassle you personally, but is there anything else I need to do here?  I have a branch for #73125 which is blocked on this.

dtolnay reviewed (approved):
Thanks for the reminder. I just had way too many things stack up, sorry for letting this drop. Looks good!

dtolnay commented:
@bors r+

bors commented:
:pushpin: Commit 4bb8425af60eb673a932c90ee8d1b5f24c13a34e has been approved by `dtolnay`

<!-- @bors r=dtolnay 4bb8425af60eb673a932c90ee8d1b5f24c13a34e -->
<!-- homu: {"type":"Approved","sha":"4bb8425af60eb673a932c90ee8d1b5f24c13a34e","approver":"dtolnay"} -->

ijackson commented:
> Thanks for the reminder. I just had way too many things stack up, sorry for letting this drop. Looks good!

No problem, I know the feeling.  Thank you :-).

JohnTitor commented:
Failed in rollup: https://github.com/rust-lang/rust/pull/82749#issuecomment-790525956
@bors r-

ijackson commented:
> Failed in rollup: [#82749 (comment)](https://github.com/rust-lang/rust/pull/82749#issuecomment-790525956)

Sorry about that.  I will restrict the failing test cases to Linux.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2030997026) [(plain)](https://github.com/rust-lang/rust/commit/a240ff5a7713f744755855c507eab327f90cd824/checks/2030997026/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Step 5/10 : RUN npm install es-check -g
 ---> Running in 6fc05ac863be
/node-v14.4.0-linux-x64/bin/es-check -> /node-v14.4.0-linux-x64/lib/node_modules/es-check/index.js

> spawn-sync@1.0.15 postinstall /node-v14.4.0-linux-x64/lib/node_modules/es-check/node_modules/spawn-sync
> node postinstall
+ es-check@5.2.3
added 95 packages from 44 contributors in 4.308s
Removing intermediate container 6fc05ac863be
 ---> 37d05c3c96c6
---
Successfully built 8f2f8a335102
Successfully tagged rust-ci:latest
Built container sha256:8f2f8a335102d35d71adf2e4a67f999d10ff8cb0ff613c2799163a4fe7426b95
Uploading finished image to https://ci-caches.rust-lang.org/docker/9f2a38e35a8211f9c2c342213b6dabbf1ce1e957c3f9f4a6874af054aa93d446d1c6f8252277cb4118d76ddf7862eec7c972b385df9acf97d8b518b20c0181e6
upload failed: - to s3://rust-lang-ci-sccache2/docker/9f2a38e35a8211f9c2c342213b6dabbf1ce1e957c3f9f4a6874af054aa93d446d1c6f8252277cb4118d76ddf7862eec7c972b385df9acf97d8b518b20c0181e6 Unable to locate credentials
[CI_JOB_NAME=mingw-check]
---
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
skip untracked path cpu-usage.csv during rustfmt invocations
skip untracked path src/doc/book/ during rustfmt invocations
skip untracked path src/doc/rust-by-example/ during rustfmt invocations
skip untracked path src/llvm-project/ during rustfmt invocations
Diff in /checkout/library/std/src/sys/unix/process/process_unix/tests.rs at line 16:
     //   https://github.com/rust-lang/rust/pull/82749#issuecomment-790525956
     // The purpose of this test is to test our string formatting, not our understanding of the wait
     // status magic numbers.  So restrict these to Linux.
-    #[cfg(target_os = "linux")] t(0x0137f, "stopped (not terminated) by signal: 19");
-    #[cfg(target_os = "linux")] t(0x0ffff, "continued (WIFCONTINUED)");
+    #[cfg(target_os = "linux")]
+    t(0x0137f, "stopped (not terminated) by signal: 19");
+    #[cfg(target_os = "linux")]
+    t(0x0ffff, "continued (WIFCONTINUED)");
 
     // Testing "unrecognised wait status" is hard because the wait.h macros typically
     // assume that the value came from wait and isn't mad.  With the glibc I have here
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2018" "--unstable-features" "--skip-children" "--check" "/checkout/library/std/src/sys/wasm/futex_atomics.rs" "/checkout/library/std/src/sys/unix/process/process_unix.rs" "/checkout/library/std/src/sys/wasm/condvar_atomics.rs" "/checkout/library/std/src/sys/unix/process/process_unix/tests.rs" "/checkout/library/std/src/sys/wasm/args.rs" "/checkout/library/std/src/sys/wasm/mutex_atomics.rs" "/checkout/library/std/src/sys/wasm/env.rs" "/checkout/library/std/src/sys/unix/process/zircon.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
Build completed unsuccessfully in 0:00:14
```

</details>

ijackson commented:
OK I think I have fixed this now.

@rustbot modify labels -S-waiting-on-author +S-waiting-on-review

ijackson commented:
@JohnTitor Should I be asking you to re-approve it, or ...?   (Thanks for your attention and sorry for messing up your rollup.)

ijackson commented:
@JohnTitor (or anyone) Also, if I may ask: if I have a branch which I suspect might have portability problems, is there a less disruptive way for me to get the full rust CI run on it than to try to get it approved, and wait for it to make a mess for you?

I could ask the reviewer to mark it not for rollup, but ideally I would get the computers to find the problems before bothering a human reviewer with re-review over portability fixes.

JohnTitor commented:
Seems reasonable to me but I'm not part of the reviewers so r? @dtolnay 
Btw, don't worry about the rollup failure, it's not so hard to re-create a new one :)

> Also, if I may ask: if I have a branch which I suspect might have portability problems, is there a less disruptive way for me to get the full rust CI run on it than to try to get it approved, and wait for it to make a mess for you?

If it makes sense, we can run a try build (via `bors try`) to check the change on some more builders than PR's.

dtolnay commented on library/std/src/sys/unix/process/process_unix/tests.rs:
```diff
@@ -0,0 +1,29 @@
+    //   https://github.com/rust-lang/rust/pull/82749#issuecomment-790525956
+    // The purpose of this test is to test our string formatting, not our understanding of the wait
+    // status magic numbers.  So restrict these to Linux.
+    #[cfg(target_os = "linux")]
+    t(0x0137f, "stopped (not terminated) by signal: 19");
+    #[cfg(target_os = "linux")]
+    t(0x0ffff, "continued (WIFCONTINUED)");
```
For tests that we only want to *run* on some platforms, but are capable of *compiling* perfectly fine on any platform, I would prefer to do the following instead. This makes things easier for someone who needs to refactor this code on a non-Linux host later. Completely conditional-compiling out the code would hide type errors or missing arguments and it commonly leads to needless round-trips through bors to get things right, which is a frustrating experience.

Using #[cfg] would typically only be if the compilation itself is platform-specific such as one of the std::os extension traits, or if for some reason there is immense cost in compiling the tests (which is not the case here).

```suggestion
    if cfg!(target_os = "linux") {
        t(0x0137f, "stopped (not terminated) by signal: 19");
        t(0x0ffff, "continued (WIFCONTINUED)");
    }
```

ijackson replied:
> For tests that we only want to _run_ on some platforms, but are capable of _compiling_ perfectly fine on any platform, I would prefer to do the following instead. This makes things easier for someone who needs to refactor this code on a non-Linux host later. Completely conditional-compiling out the code would hide type errors or missing arguments and it commonly leads to needless round-trips through bors to get things right, which is a frustrating experience.

Thanks.  That is a very good point.

dtolnay commented on library/std/src/sys/unix/process/process_unix/tests.rs:
```diff
@@ -0,0 +1,29 @@
+    // Testing "unrecognised wait status" is hard because the wait.h macros typically
+    // assume that the value came from wait and isn't mad.  With the glibc I have here
+    // this works:
+    #[cfg(all(target_os = "linux", target_env = "gnu"))]
+    t(0x000ff, "unrecognised wait status: 255 0xff");
```
```suggestion
    if cfg!(all(target_os = "linux", target_env = "gnu")) {
        t(0x000ff, "unrecognised wait status: 255 0xff");
    }
```

dtolnay commented:
@bors r+

bors commented:
:pushpin: Commit 11ca64401a5d562898e8b5f46bd36d6d1c6dc3ef has been approved by `dtolnay`

<!-- @bors r=dtolnay 11ca64401a5d562898e8b5f46bd36d6d1c6dc3ef -->
<!-- homu: {"type":"Approved","sha":"11ca64401a5d562898e8b5f46bd36d6d1c6dc3ef","approver":"dtolnay"} -->
